segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/bookkeeper/ConcurrencyManager.java

```diff
@@ -0,0 +1,196 @@
+class ConcurrencyManager {
+    //region Members
+
+    @VisibleForTesting
+    static final double SIGNIFICANT_DIFFERENCE = 0.25;
+    @VisibleForTesting
+    static final long MIN_FREQUENCY_MILLIS = 1000;
+    @VisibleForTesting
+    static final long STALE_MILLIS = MIN_FREQUENCY_MILLIS * 10;
+    @VisibleForTesting
+    static final int MAX_STAGNATION_AGE = 20;
+    private final int minParallelism;
+    private final int maxParallelism;
+    private final Supplier<Long> timeSupplier;
+    @GuardedBy("this")
+    private long recentTotalWrittenLength;
+    @GuardedBy("this")
+    private int recentWriteCount;
+    private final AtomicReference<Snapshot> lastSnapshot;
+
+    //endregion
+
+    //region Constructor
+
+    /**
+     * Creates a new instance of the ConcurrencyManager class.
+     *
+     * @param minParallelism The minimum degree of parallelism desired.
+     * @param maxParallelism The maximum degree of parallelism desired.
+     */
+    ConcurrencyManager(int minParallelism, int maxParallelism) {
+        this(minParallelism, maxParallelism, System::nanoTime);
+    }
+
+    /**
+     * Creates a new instance of the ConcurrencyManager class.
+     *
+     * @param minParallelism The minimum degree of parallelism desired.
+     * @param maxParallelism The maximum degree of parallelism desired.
+     * @param timeSupplier   A Supplier that returns the current time, in nanoseconds.
+     */
+    @VisibleForTesting
+    ConcurrencyManager(int minParallelism, int maxParallelism, Supplier<Long> timeSupplier) {
+        this.timeSupplier = Preconditions.checkNotNull(timeSupplier, "timeSupplier");
+        Preconditions.checkArgument(minParallelism <= maxParallelism, "minParallelism must be <= maxParallelism.");
+        this.minParallelism = minParallelism;
+        this.maxParallelism = maxParallelism;
+        this.recentTotalWrittenLength = 0;
+        this.recentWriteCount = 0;
+        this.lastSnapshot = new AtomicReference<>();
+        resetSnapshot();
+    }
+
+    //endregion
+
+    //region Operations
+
+    /**
+     * Records the fact that a write with the given length has just completed successfully.
+     *
+     * @param writeLength The length of the write that completed.
+     */
+    void writeCompleted(int writeLength) {
+        synchronized (this) {
+            this.recentWriteCount++;
+            this.recentTotalWrittenLength += writeLength;
+        }
+    }
+
+    /**
+     * Gets a value indicating the current (most recent) degree of parallelism.
+     *
+     * @return The result
+     */
+    int getCurrentParallelism() {
+        return this.lastSnapshot.get().parallelism;
+    }
+
+    /**
+     * Updates the degree of parallelism (if needed), based on information collected since the last time this method was called.
+     * Rules:
+     * * Only changes something if throughput changed significantly since the last time.
+     * * Changes the degree of parallelism in the same direction as throughput.
```

Comment:
This is an odd aspect to the algorithm. If we decreased the parallelism and fill rate and throughput both went up, this will react by increasing the parallelism. Maybe that is the correct thing to do, but it seems more likely the opposite is correct. This will at the very least cause the algorithm to bias high.

Reply:
Well we don't really decrease the fill rate intentionally. It too is a function of throughput and incoming data. The higher the throughput, the more likely that we are able to process writes quicker and hence not be able to fill up the entire buffer. It's because of this that I left the fill rate out of any calculations (same with latency), since both are dependent on throughput, which is a variable used in our decision making.